cpp/arcticdb/storage/azure/azure_storage-inl.hpp

```diff
@@ -0,0 +1,340 @@
+                            write_size);
+
+            ARCTICDB_SUBSAMPLE(AzureStorageUploadObject, 0)
+            try{
+                auto blob_client = container_client.GetBlockBlobClient(blob_name);
+                blob_client.UploadFrom(dst, write_size, upload_option, get_context(request_timeout));
+            }
+            catch (const Azure::Core::RequestFailedException& e){
+                util::raise_rte("Failed to write azure with key '{}' {} {}: {}",
+                                    k,
+                                    blob_name,
+                                    static_cast<int>(e.StatusCode),
+                                    e.ReasonPhrase);
+            }
+
+            ARCTICDB_RUNTIME_DEBUG(log::storage(), "Wrote key {}: {}, with {} bytes of data",
+                                variant_key_type(k),
+                                variant_key_view(k),
+                                seg.total_segment_size(hdr_size));
+        }
+    });
+}
+
+template<class KeyBucketizer>
+void do_update_impl(
+        Composite<KeySegmentPair>&& kvs,
+        const std::string& root_folder,
+        Azure::Storage::Blobs::BlobContainerClient& container_client,
+        KeyBucketizer&& bucketizer,
+        const Azure::Storage::Blobs::UploadBlockBlobFromOptions& upload_option,
+        unsigned int request_timeout) {
+    // azure updates the key if it already exists
+    do_write_impl(std::move(kvs), root_folder, container_client, std::move(bucketizer), upload_option, request_timeout);
+}
+
+struct UnexpectedAzureErrorException : public std::exception {};
+
+template<class Visitor, class KeyBucketizer>
+void do_read_impl(Composite<VariantKey> && ks,
+                Visitor&& visitor,
+                const std::string& root_folder,
+                Azure::Storage::Blobs::BlobContainerClient& container_client,
+                KeyBucketizer&& bucketizer,
+                ReadKeyOpts opts,
+                const Azure::Storage::Blobs::DownloadBlobToOptions& download_option,
+                unsigned int request_timeout) {
+    ARCTICDB_SAMPLE(AzureStorageRead, 0)
+    auto fmt_db = [](auto&& k) { return variant_key_type(k); };
+    std::vector<VariantKey> failed_reads;
+
+    (fg::from(ks.as_range()) | fg::move | fg::groupBy(fmt_db)).foreach(
+        [&container_client, &root_folder, b=std::move(bucketizer), &visitor, &failed_reads,
+         opts=opts, &download_option, &request_timeout] (auto&& group) {
+        for (auto& k : group.values()) {
+            auto key_type_folder = object_store_utils::key_type_folder(root_folder, variant_key_type(k));
+            auto blob_name = object_store_utils::object_path(b.bucketize(key_type_folder, k), k);
+            try{
+                auto blob_client = container_client.GetBlockBlobClient(blob_name);
+                auto properties = blob_client.GetProperties(Azure::Storage::Blobs::GetBlobPropertiesOptions{}, get_context(request_timeout)).Value;
```

Comment:
We can't really make a separate request just to allocate a buffer...

Reply:
Sadly because of the limitation by the SDK, we have to bite either of below two bullets:
1. 2 http requests (one for the size of the blob to allocate the buffer, one for the download)
2. Extra copy of object data